File: framework/backend/python/packages/k1s0-rate-limit/src/k1s0_rate_limit/token_bucket.py
"""Token bucket rate limiter implementation."""

from __future__ import annotations

import asyncio
import time
from datetime import timedelta

from k1s0_rate_limit.config import TokenBucketConfig
from k1s0_rate_limit.rate_limiter import RateLimiter, RateLimitStats


class TokenBucket(RateLimiter):
    """Rate limiter using the token bucket algorithm.

    Tokens are added at a constant ``refill_rate`` up to ``capacity``.
    Each call to :meth:`try_acquire` consumes one token if available.

    Args:
        config: Token bucket configuration.

    Example::

        bucket = TokenBucket(TokenBucketConfig(capacity=10, refill_rate=1.0))
        if await bucket.try_acquire():
            process_request()
    """

    def __init__(self, config: TokenBucketConfig | None = None) -> None:
        cfg = config or TokenBucketConfig()
        self._capacity = cfg.capacity
        self._refill_rate = cfg.refill_rate
        self._tokens = float(cfg.capacity)
        self._last_refill = time.monotonic()
        self._lock = asyncio.Lock()
        self._allowed = 0
        self._rejected = 0

    def _refill(self) -> None:
        """Refill tokens based on elapsed time since last refill."""
        now = time.monotonic()
        elapsed = now - self._last_refill
        self._tokens = min(self._capacity, self._tokens + elapsed * self._refill_rate)
        self._last_refill = now

    async def try_acquire(self) -> bool:
        """Attempt to consume one token from the bucket.

        Returns:
            ``True`` if a token was available and consumed, ``False`` otherwise.
        """
        async with self._lock:
            self._refill()
            if self._tokens >= 1.0:
                self._tokens -= 1.0
                self._allowed += 1
                return True
            self._rejected += 1
            return False

    def time_until_available(self) -> timedelta:
        """Return estimated wait time until one token is available.

        Returns:
            A ``timedelta`` representing the wait duration.
        """
        if self._tokens >= 1.0:
            return timedelta()
        if self._refill_rate <= 0.0:
            return timedelta.max
        deficit = 1.0 - self._tokens
        return timedelta(seconds=deficit / self._refill_rate)

    def available_tokens(self) -> int:
        """Return the current integer number of available tokens.

        Returns:
            Non-negative integer count of whole tokens available.
        """
        return int(self._tokens)

    def stats(self) -> RateLimitStats:
        """Return a snapshot of current statistics.

        Returns:
            Rate limit statistics including allowed, rejected, and available.
        """
        return RateLimitStats(
            allowed=self._allowed,
            rejected=self._rejected,
            total=self._allowed + self._rejected,
            available=int(self._tokens),
        )

File: framework/backend/go/k1s0-rate-limit/token_bucket.go
package k1s0ratelimit

import (
	"context"
	"sync"
	"sync/atomic"
	"time"
)

// TokenBucket implements the token bucket rate limiting algorithm.
//
// Tokens are added at a constant refill rate up to a maximum capacity.
// Each request consumes one token. If no tokens are available, the request is rejected.
type TokenBucket struct {
	config     *TokenBucketConfig
	tokens     int64
	lastRefill int64 // unix nanoseconds
	allowed    int64
	rejected   int64
	total      int64
	mu         sync.Mutex
}

// NewTokenBucket creates a new TokenBucket with the given configuration.
//
// Example:
//
//	config := &k1s0ratelimit.TokenBucketConfig{
//	    Capacity:   100,
//	    RefillRate: 10.0,
//	}
//	limiter := k1s0ratelimit.NewTokenBucket(config)
func NewTokenBucket(config *TokenBucketConfig) *TokenBucket {
	config = config.Validate()
	now := time.Now().UnixNano()
	return &TokenBucket{
		config:     config,
		tokens:     config.Capacity,
		lastRefill: now,
	}
}

// refill adds tokens based on elapsed time since last refill.
// Must be called with mu held.
func (tb *TokenBucket) refill() {
	now := time.Now().UnixNano()
	last := tb.lastRefill
	elapsed := time.Duration(now - last)

	tokensToAdd := int64(elapsed.Seconds() * tb.config.RefillRate)
	if tokensToAdd <= 0 {
		return
	}

	tb.lastRefill = now
	newTokens := tb.tokens + tokensToAdd
	if newTokens > tb.config.Capacity {
		newTokens = tb.config.Capacity
	}
	tb.tokens = newTokens
}

// TryAcquire attempts to acquire a token. Returns nil if allowed, error if rejected.
func (tb *TokenBucket) TryAcquire(ctx context.Context) error {
	if ctx.Err() != nil {
		return ctx.Err()
	}

	atomic.AddInt64(&tb.total, 1)

	tb.mu.Lock()
	tb.refill()

	if tb.tokens > 0 {
		tb.tokens--
		tb.mu.Unlock()
		atomic.AddInt64(&tb.allowed, 1)
		return nil
	}

	tb.mu.Unlock()
	atomic.AddInt64(&tb.rejected, 1)
	return &ErrRateLimitExceeded{
		RetryAfter: tb.TimeUntilAvailable(),
	}
}

// TimeUntilAvailable returns the duration until the next token is available.
func (tb *TokenBucket) TimeUntilAvailable() time.Duration {
	tb.mu.Lock()
	tokens := tb.tokens
	tb.mu.Unlock()

	if tokens > 0 {
		return 0
	}

	// Time for one token to be refilled
	return time.Duration(float64(time.Second) / tb.config.RefillRate)
}

// AvailableTokens returns the current number of available tokens.
func (tb *TokenBucket) AvailableTokens() int64 {
	tb.mu.Lock()
	tb.refill()
	tokens := tb.tokens
	tb.mu.Unlock()
	return tokens
}

// Stats returns the current rate limiter statistics.
func (tb *TokenBucket) Stats() Stats {
	return Stats{
		Allowed:   atomic.LoadInt64(&tb.allowed),
		Rejected:  atomic.LoadInt64(&tb.rejected),
		Total:     atomic.LoadInt64(&tb.total),
		Available: tb.AvailableTokens(),
	}
}
